frame/contracts/src/wasm/runtime.rs

```diff
@@ -936,243 +937,274 @@ where
-	//
-	// # Return Value
-	//
-	// Returns the size of the pre-existing value at the specified key if any. Otherwise
-	// `SENTINEL` is returned as a sentinel value.
-	[seal1] seal_set_storage(ctx, key_ptr: u32, value_ptr: u32, value_len: u32) -> u32 => {
+	/// Set the value at the given key in the contract storage.
+	///
+	/// This version is to be used with a fixed sized storage key. For runtimes supporting
+	/// transparent hashing, please use the newer version of this function.
+	///
+	/// The value length must not exceed the maximum defined by the contracts module parameters.
+	/// Specifying a `value_len` of zero will store an empty value.
+	///
+	/// # Parameters
+	///
+	/// - `key_ptr`: pointer into the linear memory where the location to store the value is placed.
+	/// - `value_ptr`: pointer into the linear memory where the value to set is placed.
+	/// - `value_len`: the length of the value in bytes.
+	///
+	/// # Return Value
+	///
+	/// Returns the size of the pre-existing value at the specified key if any. Otherwise
+	/// `SENTINEL` is returned as a sentinel value.
+	#[version(1)]
+	fn seal_set_storage(ctx: Runtime<E: Ext>, key_ptr: u32, value_ptr: u32, value_len: u32) -> u32 {
 		ctx.set_storage(KeyType::Fix, key_ptr, value_ptr, value_len)
-	},
+	}
 
-	// Set the value at the given key in the contract storage.
-	//
-	// The key and value lengths must not exceed the maximums defined by the contracts module parameters.
-	// Specifying a `value_len` of zero will store an empty value.
-	//
-	// # Parameters
-	//
-	// - `key_ptr`: pointer into the linear memory where the location to store the value is placed.
-	// - `key_len`: the length of the key in bytes.
-	// - `value_ptr`: pointer into the linear memory where the value to set is placed.
-	// - `value_len`: the length of the value in bytes.
-	//
-	// # Return Value
-	//
-	// Returns the size of the pre-existing value at the specified key if any. Otherwise
-	// `SENTINEL` is returned as a sentinel value.
-	[__unstable__] seal_set_storage(ctx, key_ptr: u32, key_len: u32, value_ptr: u32, value_len: u32) -> u32 => {
+	/// Set the value at the given key in the contract storage.
+	///
+	/// The key and value lengths must not exceed the maximums defined by the contracts module
+	/// parameters. Specifying a `value_len` of zero will store an empty value.
+	///
+	/// # Parameters
+	///
+	/// - `key_ptr`: pointer into the linear memory where the location to store the value is placed.
+	/// - `key_len`: the length of the key in bytes.
+	/// - `value_ptr`: pointer into the linear memory where the value to set is placed.
+	/// - `value_len`: the length of the value in bytes.
+	///
+	/// # Return Value
+	///
+	/// Returns the size of the pre-existing value at the specified key if any. Otherwise
+	/// `SENTINEL` is returned as a sentinel value.
+	#[unstable]
+	fn seal_set_storage(
+		ctx: Runtime<E: Ext>,
+		key_ptr: u32,
+		key_len: u32,
+		value_ptr: u32,
+		value_len: u32,
+	) -> u32 {
```

Comment:
Same argument about less magic is true for the return type. We should use the return type here as seen by the function. Which is essentially:
```rust
Result<impl ConvertibleToWasm, TrapReason>
```

This is of course no valid Rust code and we expect users to write out the concrete type. Some examples how to change from status quo:
```Rust
u32 -> Result<u32, TrapReason>
() -> Result<(), TrapReason>
ReturnCode -> Result<ReturnCode, TrapReason>
```

During the parsing phase we should only accept the right hand sides type from above.